### PR TITLE
Add a ScoreManager + score-related tables to the database

### PR DESCRIPTION
**Pull request by smoogipoo**
Imports scores when dragging in.

### CODE REVIEW

**Review by Aergwyn — COMMENTED** [review]
Slowly scrolled through this while waiting at work so other than style not much I could look at. Got excited for scores in lazer now!

File: osu.Game/Beatmaps/BeatmapInfo.cs [hunk 1]
```diff
@@ -29,7 +28,6 @@ public class BeatmapInfo : IEquatable<BeatmapInfo>, IJsonSerializable, IHasPrima
             set => onlineBeatmapID = value > 0 ? value : null;
         }
 
-        [JsonIgnore]
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapInfo.cs]: Revert

File: osu.Game/Scoring/ScoreInfo.cs [hunk 2]
```diff
@@ -36,27 +35,49 @@ public class ScoreInfo : IHasFiles<ScoreFileInfo>, IHasPrimaryKey, ISoftDelete
...
+                // we potentially can't update this yet due to Ruleset being late-bound, so instead update on read as necessary.
+                mods = null;
+            }
+        }
 
-        [NotMapped, JsonIgnore]
         public User User;
 
+        [Column("User")]
         public string UserString
```
Comment (smoogipoo) [inline osu.Game/Scoring/ScoreInfo.cs]: This should be named `UserJson` to be consistent with everything else in this file. Requires new DB migration

File: osu.Game/Online/API/Requests/Responses/APIScoreInfo.cs [hunk 3]
```diff
@@ -98,21 +95,33 @@ private BeatmapMetadata metadata
...
-            Ruleset = ruleset;
+            get => base.Ruleset;
+            set
+            {
+                base.Ruleset = value;
 
-            // Evaluate the mod string
-            Mods = Ruleset.CreateInstance().GetAllMods().Where(mod => modStrings.Contains(mod.ShortenedName)).ToArray();
+                // Evaluate the mod string
+                Mods = Ruleset.CreateInstance().GetAllMods().Where(mod => modStrings.Contains(mod.Acronym)).ToArray();
```
Comment (smoogipoo) [inline osu.Game/Online/API/Requests/Responses/APIScoreInfo.cs]: Probably should have a check for `if (modStrings != null)` before this.

File: osu.Game/Overlays/Profile/Sections/Ranks/PaginatedScoreContainer.cs [hunk 4]
```diff
@@ -29,6 +30,8 @@ public PaginatedScoreContainer(ScoreType type, Bindable<User> user, string heade
             ItemsContainer.Direction = FillDirection.Vertical;
         }
 
+        private ScoreManager scoreManager;
+
```
Comment (smoogipoo) [inline osu.Game/Overlays/Profile/Sections/Ranks/PaginatedScoreContainer.cs]: Revert

File: osu.Game/Scoring/ScoreInfo.cs [hunk 5]
```diff
@@ -65,35 +86,40 @@ public string UserString
...
         public int BeatmapInfoID { get; set; }
 
-        public BeatmapInfo BeatmapInfo { get; set; }
+        public virtual BeatmapInfo Beatmap { get; set; }
 
         public long? OnlineScoreID { get; set; }
 
         public DateTimeOffset Date { get; set; }
 
-        [NotMapped, JsonIgnore]
```
Comment (smoogipoo) [inline osu.Game/Scoring/ScoreInfo.cs]: Not sure about this, might still want the `[JsonIgnore]` at least...?

File: osu.Game/Scoring/ScoreInfo.cs [hunk 6]
```diff
@@ -36,27 +35,49 @@ public class ScoreInfo : IHasFiles<ScoreFileInfo>, IHasPrimaryKey, ISoftDelete
...
+            set
+            {
+                modsJson = value;
+
+                // we potentially can't update this yet due to Ruleset being late-bound, so instead update on read as necessary.
+                mods = null;
+            }
+        }
 
-        [NotMapped, JsonIgnore]
```
Comment (smoogipoo) [inline osu.Game/Scoring/ScoreInfo.cs]: And here re: keeping `[JsonIgnore]`

File: osu.Game/Scoring/ScoreInfo.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Scoring/ScoreInfo.cs]: It's a string though..

Comment (smoogipoo) [reply osu.Game/Scoring/ScoreInfo.cs]: Yeah, so is `StatisticsJson` and `ModsJson`

Comment (smoogipoo) [reply osu.Game/Scoring/ScoreInfo.cs]: Oh... I misunderstood

**Comment by smoogipoo** [comment]
This looks good to me

**Review by peppy — APPROVED** [review]

**Pull request merged**